source/dynamical_systems/src/Ring.cpp

```diff
@@ -0,0 +1,240 @@
+    rotation_offset_(std::make_shared<Parameter<CartesianPose>>("rotation_offset", center)),
+    radius_(std::make_shared<Parameter<double>>("radius", radius)),
+    width_(std::make_shared<Parameter<double>>("width", width)),
+    speed_(std::make_shared<Parameter<double>>("speed", speed)),
+    field_strength_(std::make_shared<Parameter<double>>("field_strength", field_strength)),
+    normal_gain_(std::make_shared<Parameter<double>>("normal_gain", normal_gain)),
+    angular_gain_(std::make_shared<Parameter<double>>("angular_gain", angular_gain)) {
+  this->set_center(center);
+  this->set_rotation_offset(Eigen::Quaterniond::Identity());
+}
+
+CartesianState Ring::compute_dynamics(const CartesianState& state) const {
+  // put the point in the reference of the center
+  CartesianPose pose(state);
+  pose = this->get_center().inverse() * pose;
+  // apply the rotation offset
+  pose.set_orientation(this->get_rotation_offset().conjugate() * pose.get_orientation());
+
+  CartesianTwist twist(pose.get_name(), pose.get_reference_frame());
+  double localFieldStrength;
+  twist.set_linear_velocity(this->calculateLocalLinearVelocity(pose, localFieldStrength));
+  twist.set_angular_velocity(this->calculateLocalAngularVelocity(pose, twist.get_linear_velocity(), localFieldStrength));
+
+  // transform the twist back to the base reference frame
+  return CartesianTwist(this->get_center()) * twist;
+}
+
+/**
+ * Use the local position in the circle frame to calculate return the 3D linear velocity in the local frame.
+ * @return Local linear velocity
+ */
+Eigen::Vector3d Ring::calculateLocalLinearVelocity(const CartesianPose& pose,
+                                                   double& localFieldStrength) const {
+  Eigen::Vector3d localLinearVelocity = Eigen::Vector3d::Zero();
+
+  // get the 2d components of position on the XY plane
+  Eigen::Vector2d position2d(pose.get_position().x(), pose.get_position().y());
+
+  double d = position2d.norm();
+  if (d < 1e-7) {
+    return localLinearVelocity;
+  }
+
+  double re = M_PI_2 * (d - this->get_radius()) / this->get_width();
+  if (re > M_PI_2) {
+    re = M_PI_2;
+  } else if (re < -M_PI_2) {
+    re = -M_PI_2;
+  }
+
+  // calculate the velocity of a point as an orthogonal unit vector, rectified towards the radius based on re
+  Eigen::Matrix2d R;
+  R << -sin(re), -cos(re), cos(re), -sin(re);
+  Eigen::Vector2d velocity2d = R * position2d / d;
+
+  // scale by the nominal speed
+  velocity2d *= this->get_speed();
+
+  // calculate the normal velocity
+  double vz = -this->get_normal_gain() * pose.get_position().z();
+
+  // combine into 3D velocity
+  localLinearVelocity << velocity2d, vz;
+
+  // calculate the field strength and scale the velocity
+  localFieldStrength = this->get_field_strength() + (1 - this->get_field_strength()) * cos(re);
+  localLinearVelocity *= localFieldStrength;
+
+  return localLinearVelocity;
+}
+
+/**
```

Comment:
Yes true I ported this implementation from CITRIFIED, where the style guide is camelcase. Since this is a source file I thought it's probably ok, but let me know If you would prefer this style to be matched also.

Reply:
I like our style guide from LASA but I'd like consistency even more I think.

Reply:
I find snake case more readable but it is probably because I switched to python for quite a while and it is actually standard there. As c++ does not really have a standard on the question and both are kind of accepted I figured I would keep snake case. ROS2 is also using snake case now but I guess it is to have consistency between nodes written in c++ and python. Now I don't mind for the non public API to have unconsistencies but it is also nice to see consistency everywhere. I let you decide if it is not too much.